safekeeper/src/pull_timeline.rs

```diff
@@ -158,13 +162,41 @@ impl WalResidentTimeline {
     async fn start_snapshot<W: AsyncWrite + Unpin + Send>(
         &self,
         ar: &mut tokio_tar::Builder<W>,
+        source: NodeId,
+        destination: NodeId,
     ) -> Result<SnapshotContext> {
         let mut shared_state = self.write_shared_state().await;
         let wal_seg_size = shared_state.get_wal_seg_size();
 
-        let cf_path = self.get_timeline_dir().join(CONTROL_FILE_NAME);
-        let mut cf = File::open(cf_path).await?;
-        ar.append_file(CONTROL_FILE_NAME, &mut cf).await?;
+        let mut control_store = TimelinePersistentState::clone(shared_state.sk.state());
+        let replace = control_store
+            .partial_backup
+            .replace_uploaded_segment(source, destination)?;
```

Comment:
It wasn't obvious for me that this function modifies `.partial_backup` in-place, probably worth a comment.

Another option is to just explicitly reassign the state:
```
control_store.partial_backup = wal_backup_partial::State {
  segments: vec![replace.current],
};
```

And then you don't need a reference.

Reply:
Added a comment in https://github.com/neondatabase/neon/pull/8628/commits/f423f9f7d8d89bb4e5753d3f5f661372abb71a94